src/js/actions/sampler.js

```diff
@@ -261,12 +289,17 @@ define(function (require, exports) {
             }
         });
 
-        var shapePromise = shapeLayers.isEmpty() ? Promise.resolve() :
-                this.transfer(shapeActions.setStroke, doc, shapeLayers, stroke),
+        var transaction = descriptor.beginTransaction(transactionOpts),
+            actionOpts = {
+                transaction: transaction
+            },
+            shapePromise = shapeLayers.isEmpty() ? Promise.resolve() :
+                this.transfer(shapeActions.setStroke, doc, shapeLayers, stroke, actionOpts),
             textPromise = textLayers.isEmpty() ? Promise.resolve() :
-                this.transfer(typeActions.setColor, doc, textLayers, color, false, false);
+                this.transfer(typeActions.setColor, doc, textLayers, color, actionOpts),
+            transactionPromise = descriptor.endTransaction(transaction);
 
-        return Promise.join(shapePromise, textPromise);
+        return Promise.join(shapePromise, textPromise, transactionPromise);
```

Comment:
I guess in practice this may not matter, but I think you should wait for the shape and text promises to resolve before "ending" the transaction.